poppy/wfe.py

```diff
@@ -353,6 +353,225 @@ def get_opd(self, wave):
+            alpha = self.psd_parameters[n][0]
+            beta = self.psd_parameters[n][1]
+            outer_scale = self.psd_parameters[n][2]
+            inner_scale = self.psd_parameters[n][3]
+            surf_roughness = self.psd_parameters[n][4]
+            
+            # unit check
+            psd_units = beta.unit / ((dk.unit**2)**(alpha/2))
+            assert surf_roughness.unit == psd_units, "PSD parameter units are not consistent, please re-evaluate parameters."
+            surf_unit = (psd_units*(dk.unit**2))**(0.5)
+            
+            # initialize loop-internal PSD matrix
+            psd_local = np.zeros_like(psd)
+            
+            # Calculate the PSD equation denominator based on outer_scale presence
+            if outer_scale.value == 0: # skip out or else PSD explodes
+                # temporary overwrite of k_map at k=0 to stop div/0 problem
+                k_map[cen][cen] = 1*dk.value
+                # calculate PSD as normal
+                psd_denom = (k_map**2)**(alpha/2)
+                # calculate the immediate PSD value
+                psd_interm = (beta.value*np.exp(-((k_map*inner_scale)**2))/psd_denom)
+                # overwrite PSD at k=0 to be 0 instead of the original infinity
+                psd_interm[cen][cen] = 0
+                # return k_map to original state
+                k_map[cen][cen] = 0
+            else:
+                psd_denom = ((outer_scale.value**(-2)) + (k_map**2))**(alpha/2) # unitless currently
+                psd_interm = (beta.value*np.exp(-((k_map*inner_scale)**2))/psd_denom)
+            
+            # apply surface roughness
+            psd_interm = psd_interm + surf_roughness.value
+            
+            # apply as the sum with the weight of the PSD model
+            psd = psd + (self.psd_weight[n] * psd_interm) # this should all be m2 [surf_unit]2, but stay unitless for all calculations
+        
+        # set the random noise
+        psd_random = np.random.RandomState()
+        psd_random.seed(self.seed)
+        rndm_noise = np.fft.fftshift(np.fft.fft2(psd_random.normal(size=(self.screen_size, self.screen_size))))
+        
+        psd_scaled = (np.sqrt(psd/(wave.pixelscale.value**2)) * rndm_noise)
+        opd = ((np.fft.ifft2(np.fft.ifftshift(psd_scaled)).real*surf_unit).to(u.m)).value 
+        
+        # Set rms value based on the active region of beam
+        if self.rms is not None:
+            beam_diam = wave.pixelscale * u.pix * wave_size # has units, needed for circ
+            circ = CircularAperture(name='beam diameter', radius=beam_diam/2)
+            ap = circ.get_transmission(wave)
+            opd_crop = utils.pad_or_crop_to_shape(array=opd, target_shape=wave.shape)
+            active_ap = opd_crop[ap==True]
+            rms_measure = np.sqrt(np.mean(np.square(active_ap))) # measured rms from aperture
+            opd *= self.rms.to(u.m).value/rms_measure # appropriately scales entire OPD
```

Comment:
Couple questions about this part. This makes the area used in computing the RMS of this optic depend, not on this optic itself, but on the incident beam it is illuminated with. It also implicitly assumes that the illuminated beam diameter is the same as the full input wavefront (this may not be the case, if the wavefront has been stopped down by some prior aperture). 

The way this is handled for the ZernikeWFE or ParameterizedWFE classes is that they have a `radius` attribute on the WFE optic itself, which in effect defines the clear aperture radius of that optic, over which the WFE is to be evaluated. I wonder if that approach should also be used here. One benefit would be to make more consistent the API between all three of these classes, so they each have a radius parameter consistently. What do you think?

Reply:
The radius parameter makes a lot of sense. I've been so accustomed to the MagAO-X optics being consistent with their optical diameters across the board that I never thought of it that way. I also in favor of maintaining consistency, especially if it helps future users. I looked at the other classes and it's easy to implement. I'll include it with my next commit shortly. It might take a bit because I'll also need to fix the test case at the same time.